packages/dev/materials/src/custom/customMaterial.ts

```diff
@@ -132,47 +132,51 @@ export class CustomMaterial extends StandardMaterial {
-            .replace("#define CUSTOM_VERTEX_UPDATE_NORMAL", this.CustomParts.Vertex_Before_NormalUpdated ? this.CustomParts.Vertex_Before_NormalUpdated : "")
-            .replace("#define CUSTOM_VERTEX_MAIN_END", this.CustomParts.Vertex_MainEnd ? this.CustomParts.Vertex_MainEnd : "");
-
-        if (this.CustomParts.Vertex_After_WorldPosComputed) {
-            Effect.ShadersStore[name + "VertexShader"] = Effect.ShadersStore[name + "VertexShader"].replace(
-                "#define CUSTOM_VERTEX_UPDATE_WORLDPOS",
-                this.CustomParts.Vertex_After_WorldPosComputed
-            );
-        }
+        return name;
+    }
 
-        Effect.ShadersStore[name + "PixelShader"] = this.FragmentShader.replace(
-            "#define CUSTOM_FRAGMENT_BEGIN",
-            this.CustomParts.Fragment_Begin ? this.CustomParts.Fragment_Begin : ""
-        )
-            .replace("#define CUSTOM_FRAGMENT_MAIN_BEGIN", this.CustomParts.Fragment_MainBegin ? this.CustomParts.Fragment_MainBegin : "")
-            .replace(
-                "#define CUSTOM_FRAGMENT_DEFINITIONS",
-                (this._customUniform ? this._customUniform.join("\n") : "") + (this.CustomParts.Fragment_Definitions ? this.CustomParts.Fragment_Definitions : "")
-            )
-            .replace("#define CUSTOM_FRAGMENT_UPDATE_DIFFUSE", this.CustomParts.Fragment_Custom_Diffuse ? this.CustomParts.Fragment_Custom_Diffuse : "")
-            .replace("#define CUSTOM_FRAGMENT_UPDATE_ALPHA", this.CustomParts.Fragment_Custom_Alpha ? this.CustomParts.Fragment_Custom_Alpha : "")
-            .replace("#define CUSTOM_FRAGMENT_BEFORE_LIGHTS", this.CustomParts.Fragment_Before_Lights ? this.CustomParts.Fragment_Before_Lights : "")
-            .replace("#define CUSTOM_FRAGMENT_BEFORE_FRAGCOLOR", this.CustomParts.Fragment_Before_FragColor ? this.CustomParts.Fragment_Before_FragColor : "")
-            .replace("#define CUSTOM_FRAGMENT_MAIN_END", this.CustomParts.Fragment_MainEnd ? this.CustomParts.Fragment_MainEnd : "");
-
-        if (this.CustomParts.Fragment_Before_Fog) {
-            Effect.ShadersStore[name + "PixelShader"] = Effect.ShadersStore[name + "PixelShader"].replace(
-                "#define CUSTOM_FRAGMENT_BEFORE_FOG",
-                this.CustomParts.Fragment_Before_Fog
-            );
-        }
+    protected _injectCustomCode(code: string, shaderType: string): string {
+        const codeInjectionPoints = Object.keys(this._getCustomCode(shaderType) || {});
+        codeInjectionPoints.forEach((point) => {
+            const injectedCode = this._getCustomCode(shaderType)?.[point];
 
-        return name;
+            if (injectedCode && injectedCode.length > 0) {
+                const fullPointName = "#define " + point;
+                code = code.replace(fullPointName, "\n" + injectedCode + "\n" + fullPointName);
+            }
+        });
+
+        return code;
+    }
+
+    protected _getCustomCode(shaderType: string): Nullable<{ [pointName: string]: string }> {

```

Comment:
You can remove `Nullable`, shaderType can't be something other than `vertex` or `fragment`.
If you really want, you can throw an exception if `shaderType` is not one of these values.